spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,232 @@
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import scala.reflect.ClassTag
+
+
+/** Base table scan relation*/
+case object BaseScanType extends ScanType  {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],
+                            connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+                            sqlContext: SQLContext) : BaseRelationImpl =
+    new TableScanRelationImpl(table, keyspace, cluster, connector, readConf, writeConf, userSpecifiedSchema, sqlContext)
+
+}
+
+/** Pruned columns table scan relation*/
+case object PrunedScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],
+                            connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+                            sqlContext: SQLContext) : BaseRelationImpl=
+    new PrunedScanRelationImpl(table, keyspace, cluster, connector, readConf, writeConf,
+      userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and filtered table scan relation*/
+case object PrunedFilteredScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],
+                            connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+                            sqlContext: SQLContext) : BaseRelationImpl =
+    new PrunedFilteredScanRelationImpl(table, keyspace, cluster,  connector, readConf, writeConf,
+      userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and Catalyst filtered table scan relation*/
+case object CatalystScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],
+                            connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+                            sqlContext: SQLContext) : BaseRelationImpl =
+    new CatalystScanRelationImpl(table, keyspace, cluster, connector, readConf, writeConf,
+      userSpecifiedSchema, sqlContext)
+}
+
+/** Table scan type */
+sealed trait ScanType {
```

Comment:
This should be declared before concrete ScanTypes. Scala uses bottom-up order of members: always define before use.